Support writing back updated JSON manifests

wjt opened the pull request:
Some caveats:

* There's no plumbing to commit the result to Git, push a branch and open a pull request: this still has to be run by hand
* YAML output is not supported, for reasons explained briefly in the relevant commit

But it's a start, and I have used it to update endlessm/eos-google-chrome-app and flathub/com.dropbox.Client.

https://phabricator.endlessm.com/T19948

wjt committed:
Deduplicate string -> ExternalData.Type mapping

wjt committed:
checker: remove stray semicolon

wjt committed:
Store JSON source node in ExternalData objects

This will allow us to tell them to patch the JSON later.

wjt committed:
checker: use a list comprehension

I think this is clearer.

wjt committed:
Store new version data as its own structure

This avoids each checker having to laboriously clone the ExternalData.

wjt committed:
Placate flake8

wjt committed:
Add main() wrapper function

This is quite a common idiom and means that variables local to the 'if
__name__ == "__main__"' block are not in global scope.

wjt committed:
Parse --extra-data= more correctly

Previously, the code assumed that the INSTALL-SIZE field was an empty
string. This argument is defined to be 5 :-separated fields. We can
avoid accidentally splitting on an : in the URL (the trailing field) by
passing the 'maxsplit' argument to str.split().

wjt committed:
checker: keep track of all manifest contents

This will allow us to write back changes.

wjt committed:
utils: convert Content-Length header to an int

wjt committed:
Support writing back updated JSON manifests

YAML is not supported because it's surprisingly difficult to produce
output which conforms to
https://github.com/flathub/flathub/wiki/YAML-Style-Guide. The
problematic parts are:

- The style guide requires a newline between each module definition, but
  good luck teaching a generic YAML serializer to do that.

- The style guide requires child sequences to be indented:

    # GOOD:
    modules:
      - name: foo
        sources:
          - type: bar

    # BAD:
    modules:
    - name: foo
      sources:
      - type: bar

  This doesn't appear to be possible with the standard(-ish) "yaml"
  module.  There's a replacement module, "ruamel.yaml", which does seem
  flexible enough to represent this.

The pragmatic reason to not implement this is that none of the apps
which I get nagged about use YAML manifests.

https://phabricator.endlessm.com/T19948

wjt committed:
Remove --json mode

This is superseded by --update.

wjt commented on src/checker.py:
```diff
@@ -155,3 +164,13 @@ def get_outdated_external_data(self):
+    def update_manifests(self):
+        """Updates references to external data in manifests."""
+        changed = any(data.update() for data in self._external_data)
+        if changed:
```
This needs to be per-file; otherwise, unchanged files get spuriously reindented. (Found while testing on https://github.com/flathub/com.adobe.Flash-Player-Projector)

wjt committed:
Simplify discovering checkers

It's much simpler to just make a list.

wjt committed:
Compare current and new versions more thoroughly

It can't hurt!

wjt committed:
urlchecker: check remote checksum and size

https://github.com/flathub/com.adobe.Flash-Player-Projector has:

  {
    "type" : "extra-data",
    "filename" : "flash_player_sa_linux.x86_64.tar.gz",
    "url" : "https://fpdownload.macromedia.com/pub/flashplayer/updaters/32/flash_player_sa_linux.x86_64.tar.gz",
    "sha256" : "997b857438a52a3afa3c8072bca60a2ff1cacd069b17662197d17e554a18e384",
    "size" : 8503904
  },

which is annoyingly half-way between being a rotating URL (the content
changes as minor versions are released) and not (the URL changes on
major version updates). Previously this checker would assume that it is
valid, even if the content has changed.

To avoid downloading the same data twice in the case where a
rotating-url source is up-to-date, make RotatingURLChecker set the state
to VALID or BROKEN if it's sure one way or the other, and make
ManifestChecker stop once ExternalData.state becomes something other
than UNKNOWN.

joaquimrocha commented on src/lib/externaldata.py:
```diff
@@ -93,6 +94,64 @@ def to_json(self):
+        if not name:
+            name = source.get('dest-filename')
+        if not name:
+            name = os.path.basename(url)
```
You could do all these with `name = result1 or result2 or result3`, right?

wjt replied:
Yep, this looks nicer; good point.

joaquimrocha commented on src/checker.py:
```diff
@@ -86,6 +86,15 @@ def _read_manifest(self, manifest_path):
+                "Updating YAML manifests is not yet supported"
+            )
+        else:
+            return json.dumps(contents, indent=4)
```
AFAIK this may end up giving you slightly a different JSON file than the input (beyond what are the changes). e.g. comments will be stripped of, possibly also the formatting and/or sorting of elements.
These are the main reasons why I didn't implement this solution from the beginning, but even with thos eventual annoyances, it's good to see this feature in now! Kudos!

wjt replied:
It actually does preserve element ordering in dicts due to [this](https://github.com/endlessm/flatpak-external-data-checker/blob/master/src/checker.py#L63)¹, but otherwise, yes, it normalises the formatting (which I consider a feature :stuck_out_tongue:) and removes comments (which are useful, sure, but personally I can live without them in exchange for this).

¹ Actually, from Python 3.6, `dict` is ordered, so `json.dumps(json.loads(s))` never reorders keys even if you don't specify `object_pairs_hook`.

joaquimrocha replied:
Awesome! Thanks for clarifying. I didn't see an answer to my other comment, so I am keeping this unmerged and let you decide.

joaquimrocha reviewed (approved):
LGTM! I have left a small nitpick comment, so I am approving the PR but leaving it for your to merge should you want to address that.

wjt committed:
Only write back files which have changed

Previously, in a multi-file project, all files would be rewritten even
if only one file needed updates. (This is particularly problematic when
one of the files is in a submodule.)

Avoid doing this by tracking where each ExternalData came from, and only
writing back those ones.

wjt committed:
test: don't hit the real Dropbox URL

At present, it takes over 10 minutes to fetch the installer on my
system. (I think they throttle repeated downloads.)

Switch to a test service that serves a redirect to a URL of your choice
(in this case, an image on the same service).

wjt committed:
Always initialize ExternalData.checker_data
